src/components/routes: add component RoutesMap for entring routes via a map interface

maceksimon opened the pull request:
Adds map widget based on openlayers library and its Vue 3 integration.

* Install vue3-openlayers (boot file openlayers.js)
* Add RoutesMap - uses library components to enable map features

Add composables:
* useRoutesMap - controls: saving routes, calculating route length, centering map, style function for rendering route paths with vertices.
* useRoutesMapDraw - controls: drawing map features, with the undo function
* useRoutesMapVectorLayer - controls: rendering saved features on the map (without drawing)
* useGeocoding - controls: naming saved routes via the `nominatim.openstreetmap.org` API

tmszi commented on src/composables/useRoutesMap.ts:
```diff
@@ -0,0 +1,227 @@
+  provide('ol-options', options);
+
+  // constants
+  const MAX_ZOOM_CENTERING_FACTOR = 15;
```
Use camelCase const name, please.

maceksimon replied:
Updated.

tmszi commented on src/composables/useRoutesMap.ts:
```diff
@@ -0,0 +1,227 @@
+
+  const mapRef = ref<InstanceType<typeof Map.OlMap> | null>(null);
+  const center = ref<Coordinate>(fromLonLat([14.4378, 50.0755]));
+  const projection = ref('EPSG:3857');
```
EPSG should be global variable.

maceksimon replied:
Moved to global config.

tmszi commented on src/composables/useRoutesMap.ts:
```diff
@@ -0,0 +1,227 @@
+  const DEFAULT_MAP_ZOOM = 13;
+
+  const mapRef = ref<InstanceType<typeof Map.OlMap> | null>(null);
+  const center = ref<Coordinate>(fromLonLat([14.4378, 50.0755]));
```
Map center coordinate should be global variable.

maceksimon replied:
Moved to global config.

tmszi commented on src/composables/useRoutesMap.ts:
```diff
@@ -0,0 +1,227 @@
+          }
+        },
+        (error) => {
+          console.error('Geolocation error:', error);
```
User don't know about web dev console. Use modal dialog to show error message, please.

maceksimon replied:
Replaced console log with Quasar Notify api https://quasar.dev/quasar-plugins/notify#notify-api

tmszi commented on src/composables/useRoutesMap.ts:
```diff
@@ -0,0 +1,227 @@
+        },
+      );
+    } else {
+      console.error('Geolocation is not supported by this browser.');
```
Same as my comment above.

maceksimon replied:
Replaced console log with Quasar Notify api https://quasar.dev/quasar-plugins/notify#notify-api

tmszi commented on src/composables/useRoutesMap.ts:
```diff
@@ -0,0 +1,227 @@
+    styles.push(
+      new Style({
+        geometry: new MultiPoint(geometry.getCoordinates()),
+        image: new Icon({
```
Do not use emebedded SVG images.

maceksimon replied:
Replaced with import from /public assets.

tmszi commented on src/composables/useRoutesMap.ts:
```diff
@@ -0,0 +1,227 @@
+      new Style({
+        geometry: new Point(geometry.getFirstCoordinate()),
+        image: new Icon({
+          src: "data:image/svg+xml,%3Csvg xmlns='http://www.w3.org/2000/svg' width='32' height='32' viewBox='0 0 24 24'%3E%3Cpath fill='currentColor' d='M12 2a7 7 0 0 0-7 7c0 5.25 7 13 7 13s7-7.75 7-13a7 7 0 0 0-7-7m4.5 7H14v5h-4V9H7.5L12 4.5Z'/%3E%3C/svg%3E",
```
Same as my comment above.

maceksimon replied:
Replaced with import from /public assets.

tmszi commented on src/composables/useRoutesMap.ts:
```diff
@@ -0,0 +1,227 @@
+    styles.push(
+      new Style({
+        geometry: new Point(geometry.getLastCoordinate()),
+        image: new Icon({
```
Same as my comment above.

maceksimon replied:
Replaced with import from /public assets.

tmszi commented on src/composables/useRoutesMapTooltip.ts:
```diff
@@ -0,0 +1,73 @@
+    unByKey(listener);
+  };
+
+  const formatGeometryLength = (line: LineString) => {
```
Func doc.

maceksimon replied:
Added.

tmszi commented on src/components/__tests__/RoutesMapToolbar.cy.js:
```diff
@@ -0,0 +1,116 @@
+      .should('be.visible')
+      .and('have.css', 'padding', '0px');
+    cy.dataCy('add-route-avatar').should('be.visible');
+    cy.dataCy('add-route-avatar').invoke('height').should('be.equal', 32);
```
`const addRouteAvatarIconSize = 32`

Check others repeated icon size number which should be const, please.

maceksimon replied:
Refactored selectors and variables to constants.

maceksimon commented on src/composables/useRoutesMap.ts:
```diff
@@ -0,0 +1,227 @@
+  };
+
+  /**
+   * Returns a string representing the length of a route in kilometers.
```
Edited.

maceksimon commented on src/components/routes/RoutesMap.vue:
```diff
@@ -0,0 +1,390 @@
+
+    /**
+     * Called when a new path is being drawn on the map.
+     * @returns {void}
```
Fixed.

tmszi commented:
MNK tiles map service URL https://mnk-osm-tiles.dopracenakole.net/{z}/{x}/{y}.png (URL string should be global app var same as OSM URL string)

You can make simple widget (menu with two items) inside map which allow switch between base maps OSM and MNK.

tmszi commented on src/components/routes/RoutesMap.vue:
```diff
@@ -0,0 +1,385 @@
+            <ol-source-osm />
+          </ol-tile-layer>
+          <!-- Zoom controls -->
+          <ol-zoom-control zoomInLabel="➕" zoomOutLabel="➖" />
```
According [OpenLayers doc ol/control/Zoom](https://openlayers.org/en/latest/apidoc/module-ol_control_Zoom-Zoom.html) `zoomInLabel`  and `zoomOutLabel` attribute has default value.

maceksimon replied:
Good point. Using default labels.

tmszi commented on src/components/routes/RoutesMapToolbar.vue:
```diff
@@ -0,0 +1,209 @@
+          data-cy="add-route-button"
+        >
+          <q-avatar
+            size="32px"
```
`size` attribute value 32px reapet, you can const for this attr

maceksimon replied:
Using const.

tmszi commented on src/components/routes/RoutesMapToolbar.vue:
```diff
@@ -0,0 +1,209 @@
+            <q-icon
+              name="mdi-pencil-remove"
+              :color="deleteEnabled ? 'white' : 'primary'"
+              size="18px"
```
Same as my comment above https://github.com/auto-mat/ride-to-work-by-bike-frontend/pull/470/files#r1682287232

maceksimon replied:
Updated.

tmszi commented on src/components/routes/RoutesMapToolbar.vue:
```diff
@@ -0,0 +1,209 @@
+        >
+          <!-- Icon -->
+          <q-icon
+            name="sym_s_my_location"
```
Unreadable name value

tmszi replied:
ok

tmszi commented on src/composables/useGeocoding.ts:
```diff
@@ -0,0 +1,61 @@
+    const [lon, lat] = coord;
+    // TODO: update fetch method (if we will use it)
+    const response = await fetch(
+      `https://nominatim.openstreetmap.org/reverse?format=json&lon=${lon}&lat=${lat}`,
```
OSM API URL should be part of global app vars.

maceksimon replied:
Saved as a global var.

tmszi commented on src/components/routes/RoutesMap.vue:
```diff
@@ -0,0 +1,385 @@
+            <!-- List header -->
+            <q-item class="bg-primary text-white text-weight-bold text-center">
+              <q-item-section class="text-subtitle2 text-uppercase"
+                >Vaše trasy</q-item-section
```
Translate string please.

maceksimon replied:
Translated.

tmszi commented on src/components/routes/RoutesMap.vue:
```diff
@@ -0,0 +1,385 @@
+              @click="onSavedRouteClick(route)"
+            >
+              <q-item-section v-if="route['startName'] && route['endName']">
+                <div>{{ `${route['startName']} → ${route['endName']}` }}</div>
```
Better not to use Unicode symbol → here. Use normal icon.

tmszi commented on src/components/routes/RoutesMap.vue:
```diff
@@ -0,0 +1,385 @@
+      :style="{
+        borderRadius,
+        overflow: 'hidden',
+        border: '1px solid #E0E0E0',
```
We need here extra border color?

maceksimon replied:
Replaced with `grey-4`

tmszi commented:
If you activate one editing tool (draw new line) other tools should be disabled.

tmszi commented:
1. Simple tooltip above mouse cursor is still missing.
2. Missing switch widget between base maps OSM, MNK

maceksimon commented:
Status
* Added tooltip with date, transport type and route length.
* Added select for layer source.
* Updated undo feature incl. tests (undo now works when tools are switched during drawing and resets after save).

tmszi commented:
I rebased PR commit https://github.com/auto-mat/ride-to-work-by-bike-frontend/commit/8269a39b75953360aed3d3ad35c03516838f281b  and (installation JS lib packages are|starting dev server) is successful without errors.

However update to new vue3js-openlayers JS lib 11 major version, brings that some map functions stopped working.
